kbfstool/md_reset.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+
+	"github.com/keybase/kbfs/libkbfs"
+	"golang.org/x/net/context"
+)
+
+func mdResetOne(
+	ctx context.Context, config libkbfs.Config, tlfPath string,
+	dryRun bool) error {
+	handle, err := parseTLFPath(ctx, config.KBPKI(), tlfPath)
+	if err != nil {
+		return err
+	}
+
+	// This function is loosely adapted from
+	// folderBranchOps.initMDLocked.
+
+	username, uid, err := config.KBPKI().GetCurrentUserInfo(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Make sure we're a writer before doing anything else.
+	if !handle.IsWriter(uid) {
+		return libkbfs.NewWriteAccessError(
+			handle, username, handle.GetCanonicalPath())
+	}
+
+	fmt.Printf("Looking for unmerged branch...\n")
+
+	_, unmergedIRMD, err := config.MDOps().GetForHandle(
+		ctx, handle, libkbfs.Unmerged)
+	if err != nil {
+		return err
+	}
+	if unmergedIRMD != (libkbfs.ImmutableRootMetadata{}) {
+		return fmt.Errorf(
+			"%s has unmerged data; try unstaging it first",
+			tlfPath)
+	}
+
+	fmt.Printf("Getting latest metadata...\n")
+
+	_, irmd, err := config.MDOps().GetForHandle(
+		ctx, handle, libkbfs.Merged)
+	if err != nil {
+		return err
+	}
+
+	if irmd == (libkbfs.ImmutableRootMetadata{}) {
+		fmt.Printf("No TLF found for %q\n", tlfPath)
+		return nil
+	}
+
+	rootPtr := irmd.Data().Dir.BlockPointer
+	if rootPtr.Ref().IsValid() {
+		var dirBlock libkbfs.DirBlock
+		err = config.BlockOps().Get(ctx, irmd, rootPtr, &dirBlock)
+		if err == nil {
+			fmt.Printf("Got no error when getting root block %s; aborting\n", rootPtr)
+			return nil
+		}
+		fmt.Printf("Got error %s when getting root block %s, so revision %d is broken. Making successor...\n",
+			err, rootPtr, irmd.Revision())
+	} else {
+		// This happens in the wild, but only for folders used
+		// for journal-related testing early on.
+		fmt.Printf("Root block pointer is invalid, so revision %d is broken. Making successor...\n",
+			irmd.Revision())
+	}
+
+	rmdNext, err := irmd.MakeSuccessor(config.Codec(), irmd.MdID(), true)
+	if err != nil {
+		return err
+	}
+
+	_, info, readyBlockData, err :=
+		libkbfs.ResetRootBlock(ctx, config, uid, rmdNext)
```

Comment:
>As an aside, why is that? Is QR too aggressive?

Yeah this would be more useful once we allow histories to stick around.  This is fine for now, just wanted to make sure we had it in mind for the future.